### DISCUSSION TITLE
Custom plugin? Can I import recordSetStore?

### DISCUSSION THREAD

**Discussion by ECrosland** [bryanmylee/svelte-headless-table#111] [Q&A] [answered]
Can I create my own plugin? I thought I'd attempt to put one together following the patterns I can discern in the plugins that are distributed with Headless Table. However, it doesn't seem as though I am able to import things such as type `TablePlugin`, or the `recordSetStore` function, which I would like to use in my implementation.

Please forgive my ignorance if I'm overlooking something obvious. I'm new to TypeScript and JS modules in general.

I'm able to have an import statement such as the following:

`import { createTable, createRender, Subscribe, Render, } from "svelte-headless-table"`

However, I get an error if I try this:

`import { recordSetStore, type RecordSetStore } from "svelte-headless-table/utils/store"`

The error is "Could not find a declaration file for module 'svelte-headless-table/utils/store'". Is this module not exported? Or am I just not using the correct syntax in the from clause?

---

Edit: FWIW, I think I understand better now how packaging works. It wasn't just TypeScript and JS modules that I needed to understand to make sense of this error, but also the role of `config.package` in the  `svelte.config.js` file, and, ultimately, `exports` in `package.json` in the published package. The more you know.

**Comment by bryanmylee** [comment, marked as answer]
Ahh the module isn't exported. I didn't consider exporting it initially, but sure I'd be happy to update it!

**Reply by ECrosland** [reply]
Ah, OK. Thanks for the reply!

Yeah, it seems as though there are a number of types in `types/TablePlugin` that I would likely want to reference to create a plugin, plus also some things from `utils` that would be useful in the implementation (the `RecordSetStore` in the case of the plugin I have in mind). I know exporting them creates a larger public API surface, which you'd have to support going forward, but, if you'd be willing to do so, I think it would be cool for consumers of the library to be able to create custom plugins.

**Reply by ECrosland** [reply]
On the other hand, maybe I leapt too quickly to trying to create a custom plugin. It seems that the `addSelectedRows` plugin will probably be enough for me after all, especially if I can attach multiple instances of this plugin to my table to represent different kinds of selection.

So I don't know if you will want to go to the trouble of making custom plugins possible right now.

Regardless, I appreciate your reply. It helped me to figure out the piece that was missing from my understanding, which had to do with a package specifying what it exports.
